swig/cpp/src/Tree_Schema.hpp

```diff
@@ -126,6 +127,15 @@ class Submodule
 public:
     /** wrapper for struct [lys_submodule](@ref lys_submodule), for internal use only */
     Submodule(struct lys_submodule *submodule, S_Deleter deleter);
+    /** creates a Submodule from `module` if it is a submodule */
+    Submodule(S_Module module):
+        submodule((lys_submodule*)module->module),
+        deleter(module->deleter)
+    {
+        if (module->type() != 1) {
```

Comment:
Not commenting on this patch on its own, but I really dislike the way how this is done. If only that `type` member was named more descriptively, perhaps something like `is_submodule`, for example. Compare this:
```C
if (node->module->type) {
    ly_print(out, ",\"included-from\":\"%s\"", node->module->name);
}
```
with this:
```C
if (node->module->is_submodule) {
    ly_print(out, ",\"included-from\":\"%s\"", node->module->name);
}
```

Reply:
To be completely honest, I dislike a lot of things about the entire C++ bidings (especially how class are casted to each other), but I was trying to mimic what's already there - `Schema_Node_*` ctors are also defined in the header file (but no reason to mimic "bad" code I guess). About `::type`: it's called that even in the C struct, so no way of changing that, without breaking the API, but in C++, there could be a method which would do the same thing, but would be named `is_submodule`.

Reply:
Yeah, my comment was meant as feedback to libyang maintainers.

Reply:
There are a lot of things wrong about these C++ bindings mostly because they were not written by the authors but it will stay the way it is in libyang version 1.x and will be completely rewritten for libyang 2.x.